### PR TITLE
fix(proxy): prevent account bursts and implicit fallback escape

### PR DESCRIPTION
**Pull request by murdore**
## Summary
- add per-account upstream admission with a hot-reloadable `routing.max-inflight-per-account` cap (default 2), failover to a later available account, and stream-lifetime lease release
- make translation-layer auto fallback explicit (`routing.auto-fallback: true`); otherwise empty fallback chains stay inside the configured account/provider set
- rotate immediately on HTTP 529 overloads, skip unhealthy configured fallback providers, and sanitize fallback errors before durable diagnostics
- document both runtime-reloadable routing controls

## Verification
- `pnpm exec vitest run test/proxyConfigHotReload.test.ts test/proxyUpdaterFallback.test.ts test/proxyReliabilityHardening.test.ts` (127 passed)
- `pnpm exec tsc --noEmit --strict`
- `pnpm run build:cli`
- `pnpm run test:bugfixes` (234 passed)
- pre-commit hook: `check`, staged formatting, repository lint, and `prepack`

## Safety
- implemented and tested only in an isolated worktree; the live proxy was not restarted, reconfigured, or used as a test target.

## Summary by CodeRabbit

* **New Features**
  * Added configurable per-account concurrency limits, defaulting to two simultaneous requests.
  * Added FIFO queuing and account rotation for busy or overloaded upstreams.
  * Added opt-in automatic provider fallback.
  * Routing settings can be updated at runtime without restarting the proxy.
* **Bug Fixes**
  * Streaming requests release account capacity when finished or cancelled.
  * Invalid configuration values are safely ignored with warnings.
  * Provider credentials and sensitive error details are redacted from diagnostics.
* **Documentation**
  * Updated the proxy configuration reference with new routing options and validation rules.

### CODE REVIEW

**Comment by Tara-ag** [comment]
## Review Summary

This PR adds three critical improvements to the Claude proxy routing system:

### 1. **Explicit Auto-Fallback Control** (MAJOR)
- **Problem**: Previously, an empty fallback chain would silently add an "auto-provider", potentially routing requests to unexpected providers without explicit configuration
- **Fix**: `autoFallback` is now disabled by default and requires explicit opt-in via `routing.auto-fallback: true`
- **Impact**: Prevents implicit provider selection, keeps fallback behavior predictable and secure

### 2. **Per-Account Admission Control** (MINOR)
- **New feature**: `routing.max-inflight-per-account` limits concurrent upstream requests per OAuth account (default: 2, valid range: 1-20)
- **Behavior**: When all accounts reach capacity, later eligible accounts can still serve requests instead of queuing
- **Benefit**: Prevents account burst abuse while maintaining request throughput

### 3. **HTTP 529 Overload Detection** (MINOR)
- **Improvement**: Detects HTTP 529 gateway overload errors and SSE overloaded responses
- **Action**: Immediately rotates away from overloaded accounts without wasting transient retries
- **Benefit**: Faster failover during capacity constraints

### Impact on Existing Code
- ✅ No breaking changes - all new features are backward compatible with sensible defaults
- ✅ Hot-reloadable config options - no restart needed
- ✅ Type-safe additions only
- ✅ Comprehensive test coverage added for all new functionality
- ✅ Minimal blast radius - changes confined to proxy routing layer

### Code Quality Assessment
- All TypeScript types properly defined in canonical location (`src/lib/types/`)
- Validation handles edge cases (kebab-case/camelCase, string/boolean inputs)
- Lease management ensures cleanup even on stream disconnect
- Tests verify correct behavior across all scenarios

**Decision: APPROVED**

The changes are safe to merge. They address real operational concerns (burst control, overload handling) without introducing breaking changes or architectural violations.

**Review by Tara-ag — APPROVED** [review]

**Review by Tara-ag — APPROVED** [review]

**Comment by murdore** [comment]
Addressed the review findings in a37d1a13:

- Propagated explicit `routing.auto-fallback` through Claude and OpenAI translation plans.
- Strictly validate numeric admission limits and share the range/default constants.
- Bound and abort admission waiters; when all accounts are busy, reserve the first account that becomes available.
- Bound redacted fallback diagnostics and make stream lease release wait for terminal telemetry/final request logging.
- Added focused and continuous-suite regressions, including 529 rotation, admission cleanup, and stream finalization ordering.

Validated with strict TypeScript, CLI build, 131 focused tests, and 237 continuous bugfix tests. The repository-wide Prettier scan in the pre-commit hook terminates in this checkout before ESLint; changed-file formatting and focused ESLint are clean.

**Comment by Tara-ag** [comment]
## Yama Review Summary

**Decision:** CHANGES_REQUESTED

### Findings Summary:

1. **MAJOR (3 findings):**
   - Type inconsistency in optional accessor pattern (`isAutoFallbackEnabled?(): boolean` vs `.isAutoFallbackEnabled?.()`)
   - Missing JSDoc documentation for new types (AccountAdmissionState, AccountAdmissionWaiter, QueuedAccountAdmission)
   - Duplicated concurrency-bound constants across multiple files risking inconsistency

2. **MINOR (0 findings):** None

### Impact on existing code:

The changes touch proxy routing infrastructure with significant blast radius:
- **Changed files:** 13 files (core proxy logic + tests + docs)
- **Blast radius:** ~13K impacted nodes based on graph analysis
- **Critical flows:** All Claude proxy routes use ModelRouter, so any API changes affect all providers using the proxy

### Backward Compatibility Assessment:

✅ **SAFE:** The new fields are optional accessors/methods that return `undefined` when not configured. Existing callers that don't pass these parameters will continue to work unchanged. The `?? false` fallback pattern ensures safe defaults.

### Key Architecture Concerns:

1. **Default auto-fallback behavior changed:** When `auto-fallback` is not explicitly enabled, the "auto-provider" attempt is now skipped (previously it would be added by default). This could break some call sites.

2. **Constants drift risk:** Three copies of the same constants could lead to validation allowing more requests than runtime enforcement permits.

### Recommendation:

Address the MAJOR findings before merging:
1. Standardize the accessor method syntax (either as property or function call consistently)
2. Add JSDoc comments for new types  
3. Consolidate duplicate constants into single source of truth

The overall direction is sound (account admission control is needed), but the implementation needs cleanup to avoid subtle bugs.

**Review by Tara-ag — CHANGES_REQUESTED** [review]
## Yama Review Summary

**Decision:** CHANGES_REQUESTED

### Findings Summary:

1. **MAJOR (3 findings):**
   - Type inconsistency in optional accessor pattern (`isAutoFallbackEnabled?(): boolean` vs `.isAutoFallbackEnabled?.()`)
   - Missing JSDoc documentation for new types (AccountAdmissionState, AccountAdmissionWaiter, QueuedAccountAdmission)  
   - Duplicated concurrency-bound constants across multiple files risking inconsistency

2. **MINOR (0 findings):** None

### Impact on existing code:

The changes touch proxy routing infrastructure with significant blast radius:
- **Changed files:** 13 files (core proxy logic + tests + docs)
- **Blast radius:** ~13K impacted nodes based on graph analysis
- **Critical flows:** All Claude proxy routes use ModelRouter, so any API changes affect all providers using the proxy

### Backward Compatibility Assessment:

✅ **SAFE:** The new fields are optional accessors/methods that return `undefined` when not configured. Existing callers that don't pass these parameters will continue to work unchanged. The `?? false` fallback pattern ensures safe defaults.

### Key Architecture Concerns:

1. **Default auto-fallback behavior changed:** When `auto-fallback` is not explicitly enabled, the "auto-provider" attempt is now skipped (previously it would be added by default). This could break some call sites.

2. **Constants drift risk:** Three copies of the same constants could lead to validation allowing more requests than runtime enforcement permits.

### Recommendation:

Address the MAJOR findings before merging:
1. Standardize the accessor method syntax (either as property or function call consistently)
2. Add JSDoc comments for new types  
3. Consolidate duplicate constants into single source of truth

The overall direction is sound (account admission control is needed), but the implementation needs cleanup to avoid subtle bugs.

File: src/lib/proxy/modelRouter.ts [hunk 1]
```diff
@@ -41,6 +51,14 @@ export class ModelRouter {
     return this.fallback;
   }
```
Comment (Tara-ag) [inline src/lib/proxy/modelRouter.ts:52]: 💬 SUGGESTION: Consider adding JSDoc comments for newly added API surface

The newly added methods `isAutoFallbackEnabled()` (line 52) and `getMaxInflightPerAccount()` (line 56) on ModelRouter should have JSDoc documentation explaining their purpose and usage.

**Suggestion:** Add JSDoc comments before these method declarations to document the public API contract.

**Review by Tara-ag — APPROVED** [review]

**Review by Tara-ag — CHANGES_REQUESTED** [review]
## Code Review for PR #1252

This PR adds auto-fallback and max-inflight-per-account configuration to the Claude proxy routing. Key changes:

1. **ModelRouter**: Added per-account admission state with configurable concurrency limits
2. **RoutingPolicy**: Restructured fallback handling to use new admission system, immediate rotation on 529 overload
3. **ProxyConfig**: Added validation for auto-fallback and max-inflight-per-account fields
4. **Tests**: Comprehensive test coverage for new behavior

**Impact**: High - affects all Claude proxy requests with account-based load balancing

**Findings:**
- 🔒 MAJOR: HTTP 529 overload handling lacks context about active waiters (routingPolicy.ts:6478)
- 💡 MINOR: Missing JSDoc for newly added admission functions (routingPolicy.ts:6150)

**Comment by Tara-ag** [comment]
## Code Review Summary for PR #1252

### Decision: **CHANGES_REQUESTED**

This PR adds significant improvements to the Claude proxy routing with auto-fallback and max-inflight-per-account configuration. The changes are well-tested and architecturally sound, but there are several documentation gaps that need to be addressed before merge.

### Findings Summary:
- **🔒 MAJOR**: 1 issue (HTTP 529 overload handling lacks context about active waiters)
- **💡 MINOR**: 3 issues (missing JSDoc documentation)
- **💬 SUGGESTION**: 4 items (documentation improvements)

Total: 8 findings across 8 files

### Key Files Changed:
1. `src/lib/proxy/modelRouter.ts` - Added account admission state management
2. `src/lib/proxy/routingPolicy.ts` - Restructured fallback handling with new admission system
3. `src/lib/proxy/proxyConfig.ts` - Added validation for new config fields
4. `src/lib/server/routes/openaiProxyRoutes.ts` - Enhanced auth retry handling
5. `src/lib/types/proxy.ts` - Added ProxyRoutingConfig type
6. `src/lib/types/subscription.ts` - Added SubscriptionStatus enum
7. Test files with comprehensive coverage for new behavior
8. Documentation updated with hot reload notes

### Impact Analysis:
- **Blast radius**: High - affects all Claude proxy requests with account-based load balancing
- **Execution flows**: Modified authentication flow, request routing, fallback logic
- **Risk score**: 0.85 (high) due to complexity of admission state management

### Blocking Criteria Check:
✅ No hardcoded secrets or API keys
✅ No critical security vulnerabilities  
✅ No breaking changes to public SDK API
✅ No violations of CLAUDE.md Critical Rules
✅ Less than 3 MAJOR severity issues (only 1)

### Recommendation:
The PR is ready for merge after addressing the documentation gaps. The single MAJOR issue (HTTP 529 waiter notification) is a nice-to-have improvement that doesn't block the core functionality. Consider addressing it in a follow-up PR if the workload per account is expected to be high.

---
*Reviewed by Yama Code Review Agent*

**Comment by Tara-ag** [comment]
🔒 MAJOR: HTTP 529 overload handling lacks context about active waiters. When HTTP 529 is detected, the code rotates immediately without checking if there are queued waiters that should be notified. The new admission state mechanism tracks waiters but doesn't wake them up when an overload occurs. Add notification logic to wake waiting accounts when upstream becomes overloaded. (src/lib/proxy/routingPolicy.ts:6478)

**Comment by Tara-ag** [comment]
💡 MINOR: Missing JSDoc for newly added admission functions. Several new functions manage account admission state but lack documentation. This includes acquireAccountAdmission, tryAcquireAccountAdmission, getAccountAdmissionSnapshot. Add comprehensive JSDoc comments explaining the admission queue behavior, timeout handling, and abort signal usage. (src/lib/proxy/routingPolicy.ts:6150)

**Comment by Tara-ag** [comment]
💬 SUGGESTION: Missing JSDoc for newly added authentication retry handler. The new auth retry handling logic in the proxy routes lacks comprehensive documentation explaining the retry conditions, limits, and failure scenarios. Add JSDoc comments explaining the authentication retry mechanism including max retries (3), backoff strategy, and error types that trigger retries. (src/lib/server/routes/openaiProxyRoutes.ts:1250)

**Comment by Tara-ag** [comment]
💡 MINOR: Missing JSDoc documentation for ProxyRoutingConfig type. The new ProxyRoutingConfig interface that encapsulates auto-fallback and max-inflight-per-account settings lacks comprehensive documentation explaining each field's purpose, valid values, and defaults. Add comprehensive JSDoc for ProxyRoutingConfig with examples of valid configurations and explain the relationship between these fields and account admission behavior. (src/lib/types/proxy.ts:240)

**Comment by Tara-ag** [comment]
💬 SUGGESTION: Missing JSDoc for SubscriptionStatus enum values. The newly added SubscriptionStatus enum values (PROVISIONED, OVERDRAFT) lack documentation explaining their meaning and when each state is used. Add JSDoc comments to SubscriptionStatus enum with clear descriptions of each status and conditions that trigger transitions between states. (src/lib/types/subscription.ts:350)

**Comment by Tara-ag** [comment]
💬 SUGGESTION: Missing JSDoc for newly added hot reload test utilities. Several new test utility functions for testing config hot reload lack documentation explaining their usage and expected behavior. Add JSDoc comments to helper functions like updateConfigAndWaitForReload, waitForProxyUpdate with clear descriptions of parameters and return values. (test/proxyConfigHotReload.test.ts:10)

**Comment by Tara-ag** [comment]
💬 SUGGESTION: Missing JSDoc for test fixture and helper functions. Several test fixtures and helper functions lack documentation explaining their purpose and usage. Add JSDoc comments to test helpers with clear descriptions of parameters and expected behavior. (test/continuous-test-suite-bugfixes.ts:10)

**Comment by Tara-ag** [comment]
💬 SUGGESTION: Missing detailed explanation of auto-fallback behavior in configuration reference. The updated documentation mentions that changes reload without restart but doesn't clearly explain how the auto-fallback mechanism works at runtime, when it triggers, and what conditions affect it. Add a dedicated section explaining the auto-fallback mechanism: trigger conditions (upstream failures), retry logic, fallback selection criteria, and any limits on automatic retries. (docs/features/claude-proxy-config-reference.md:340)

**Comment by Tara-ag** [comment]
🔒 MINOR: AutoFallback configuration needs usage examples - The autoFallback configuration option (enabled, minProvidersRequired, maxFailsBeforeFallback, etc.) is defined in ProxyConfig but lacks concrete usage examples showing how to configure it for different scenarios. Consider adding JSDoc examples showing common autoFallback configurations like: 'autoFallback: { enabled: true, minProvidersRequired: 2 }'

**Comment by Tara-ag** [comment]
🔒 MINOR: New proxy functions lack inline JSDoc documentation - The newly exported functions `handleAnthropicNonOkResponse`, `handleAnthropicAuthRetry`, `handleAnthropicStreamingSuccessResponse`, `acquireAccountAdmission`, etc. in routingPolicy.ts lack inline JSDoc comments explaining their purpose, parameters, and return values. Consider adding JSDoc comments to these test hook functions and public API functions.

**Comment by Tara-ag** [comment]
🔒 MINOR: AutoFallback opt-in behavior needs explicit negative test - The test 'buildProxyTranslationPlan: auto-provider requires explicit opt-in' verifies that auto-fallback is disabled by default, but there's no test verifying that auto-fallback IS enabled when explicitly set to true in ProxyRoutingConfig. Consider adding a test case that explicitly sets autoFallback: true and verifies that auto-provider fallback is added to the translation plan.

**Comment by Tara-ag** [comment]
🔒 MINOR: MAX_INFLIGHT_PER_ACCOUNT default needs clear documentation - The DEFAULT_MAX_INFLIGHT_PER_ACCOUNT constant (20) is used as the default for maxInflightPerAccount, but it's not documented where this value comes from or why 20 was chosen. Consider adding a comment explaining that 20 is chosen based on typical concurrent request patterns and memory constraints per account.

**Comment by Tara-ag** [comment]
🔒 MAJOR: Error messages may leak sensitive information - When an upstream provider rejects a request due to missing/invalid credentials, the error message includes the full authorization header which may contain API keys. While the redactProviderErrorMessage function exists, its usage needs verification across all error paths. Please verify that redactProviderErrorMessage is called before logging ANY error that might contain sensitive data like 'authorization', 'api_key', 'bearer', etc.

**Comment by Tara-ag** [comment]
🔒 MINOR: Account admission lease needs error case testing - The acquireAccountAdmission function is tested but only for successful cases. There's no test verifying what happens when the account has reached its max inflight limit and a new account tries to get an admission lease. Consider adding a test case that verifies acquireAccountAdmission returns false when the account has reached its maxInflightPerAccount limit.

**Comment by Tara-ag** [comment]
🔒 MINOR: New proxy routes need OpenAPI/Swagger documentation - The new POST /api/v1/proxy/openai/account/fallback and POST /api/v1/proxy/openai/max-inflight-per-account routes are added but don't have corresponding OpenAPI documentation or inline JSDoc explaining request/response schemas. Consider adding OpenAPI documentation (OpenAPI comment blocks or inline JSDoc) for these new endpoints specifying request parameters and response formats.

**Comment by Tara-ag** [comment]
🔒 MINOR: Documentation table needs new configuration options - The docs/features/claude-proxy-config-reference.md file is updated but the table at line 341 still references the old routing.strategy option without mentioning that it now also supports 'auto' for auto-fallback behavior. Consider updating the routing strategy table to include 'auto' as a valid value and explain what auto-fallback does.

**Comment by Tara-ag** [comment]
🔒 SUGGESTION: Max-inflight-per-account section needs usage examples - The documentation mentions --max-inflight-per-account but doesn't provide example commands showing how to use this flag or what values are acceptable. Consider adding example commands like: '--max-inflight-per-account 10' and explain the range of acceptable values.

**Comment by Tara-ag** [comment]
🔒 MINOR: Circuit breaker timeout needs explicit test - The circuit breaker mechanism is implemented but there's no test verifying that a provider is correctly marked as unavailable after the circuit opens and then recovers after the timeout period. Consider adding a test that verifies the circuit breaker opens when failures exceed threshold, waits for timeout, and then closes to allow retries.

**Comment by Tara-ag** [comment]
🔒 SUGGESTION: High concurrency scenario needs performance testing - The proxy handles multiple concurrent requests per account with admission leases, but there's no stress test verifying the system can handle high concurrency without deadlocking or race conditions. Consider adding a stress test that sends many concurrent requests from the same account to verify the admission lease mechanism prevents oversubscription.

**Comment by Tara-ag** [comment]
🔒 MINOR: New subscription types need JSDoc comments - The Subscription and related types in src/lib/types/subscription.ts are modified but lack comprehensive JSDoc documentation explaining field purposes and constraints. Consider adding JSDoc comments to Subscription, SubscriptionId, and related types explaining each field's purpose and validation rules.

**Comment by Tara-ag** [comment]
🔒 SUGGESTION: Subscription update needs idempotency test - The subscription update functionality is tested but there's no test verifying that updating a subscription with the same data as before is idempotent (no unnecessary updates). Consider adding a test that verifies calling updateSubscription twice with identical data results in only one database write.

**Comment by Tara-ag** [comment]
## 🛡️ Yama Review Verdict: BLOCKED

Severity counts — 🔒 CRITICAL: 1 · ⚠️ MAJOR: 1 · 💡 MINOR: 1

PR adds HTTP 529 overload handling with account-level admission control, fallback routing, and circuit breaker. Overall design is sound but there are several issues requiring fixes: (1) Error messages in src/lib/proxy/routingPolicy.ts may leak sensitive information (CRITICAL), (2) Consolidate duplicated concurrency-bound constants into one source (MAJOR), (3) Missing JSDoc documentation for newly added API surface (MINOR). Impact analysis shows changes affect proxy routing flow with moderate blast radius.

### Findings behind this verdict
- 🔒 **CRITICAL**: Error messages may leak sensitive information — `src/lib/proxy/routingPolicy.ts:6908`
  Error message at line 6908 includes 'account' which could be used to identify specific accounts under quota limits. This violates security best practices for error messaging.
- ⚠️ **MAJOR**: Consolidate duplicated concurrency-bound constants into one source — `src/lib/proxy/modelRouter.ts:8`
  Constants MAX_INFLIGHT_PER_ACCOUNT, MAX_WAITERS_PER_ACCOUNT, and related values are defined in multiple files (modelRouter.ts, routingPolicy.ts, etc.). This creates inconsistency risks and maintenance burden.
- 💡 **MINOR**: Missing JSDoc documentation for newly added API surface — `src/lib/proxy/modelRouter.ts:48`
  Newly added functions and types lack proper JSDoc comments explaining their purpose, parameters, and return values.

**Comment by murdore** [comment]
## Review-resolution audit

Rechecked every existing inline thread and the current branch after the rebase.

All actionable review findings have been addressed in `0fd9e8c5`, and every inline thread is resolved:

- per-account admission defaults and bounds are defined once in `modelRouter.ts` and imported by config parsing and route enforcement;
- `max-inflight-per-account` accepts only integer numeric values in the documented range;
- provider fallback diagnostics are redacted and capped at 500 characters, with focused coverage;
- auto-provider fallback is disabled by default, enabled only by explicit `routing.auto-fallback: true`, and covered in the continuous suite;
- account-admission waiters are bounded, abortable, removed on timeout/abort, and select the first account that becomes available;
- the stream lease releases after final request telemetry, with ordering coverage;
- HTTP 529 / `overloaded_error` is classified and rotates immediately, with focused coverage.

The remaining Yama `CHANGES_REQUESTED` state is not backed by an unresolved current thread. Its reported 529 wake-waiters concern is not an additional required notification path: waiters are drained when a lease is released, while a 529 immediately rotates the active request. Its other broad claims either refer to unrelated or nonexistent routes and types, a nonexistent object-shaped auto-fallback configuration, or an incorrect default of 20 (the actual shared default is 2).

Current branch state: one commit on current `release` (`65b27b75`, v10.8.5), so no further rebase is needed. CI is green except the Yama verdict-enforcement job, which is carrying the stale review decision; its review step completed and produced no newer inline finding.
